west: Update west boards formatter and fish shell completion

Emplis opened the pull request:
This PR is a follow up to that discussion: https://github.com/zephyrproject-rtos/zephyr/pull/69336#discussion_r1500512829.
Its adding the `vendor` boards formatter argument and update the board completion for fish shell as described in the comment. 

That fixes #69387.

This PR also fix an issue described in that PR https://github.com/zephyrproject-rtos/zephyr/pull/69336 where the board completion is extremely slow (several seconds). To resolve that issue, the second commit introduce a simple boards caching system. With it, displaying the board is nearly instantaneous.

Emplis committed:
west: Update fish completion for `boards`

With the recent change to hwmv2, loading the
boards is extremely slow and can take several
seconds.

To solve that problem this commit add a cache
of the boards. The cache is updated based on:
    - when the latest commit of the manifest
    file directory is updated or;
    - when the manifest file directory is not
    a git directory, when the manifest file
    content itself is updated.

At the same time:
    - update how the board completion is
    displayed by including the board vendor to it;
    - add missing `--board` and `--board-dir`
    options;
    - remove `hwmv1` board completion code.

Signed-off-by: Théo Battrel <theo.util@protonmail.ch>

Emplis committed:
west: add vendor to boards format

Add the new `vendor` formatter argument to display board vendor.

Signed-off-by: Théo Battrel <theo.util@protonmail.ch>

tejlmand commented on scripts/west_commands/completion/west-completion.fish:
```diff
@@ -196,21 +196,52 @@ function __zephyr_west_complete_help
+    set -l one_week_in_s 604800
+    set -l is_cache_valid 0
+
+    set -l cache_folder $HOME/.cache/west
```
I don't think `$HOME/.cache` is appropriate for Mac users.

Perhaps a folder under `$__fish_user_data_dir` is a better choice for fish users across OS'es.
Or the temp dir, which would be cleared on reboot.

nordicjm replied:
note: temp is not cleared on reboot on windows

theob-pro replied:
Indeed it's a problem if it doesn't work on other OS'es. But I am not sure that using `$__fish_user_data_dir` is the correct way to do. It's not well documented but it seems to be a directory used to load third-party software fish functions.
Here is what the documentation say:
> Directories for other software to put their own functions. These are in the directories under $__fish_user_data_dir (usually ~/.local/share/fish, controlled by the XDG_DATA_HOME environment variable) and in the XDG_DATA_DIRS environment variable, in a subdirectory called fish/vendor_functions.d.

Instead I propose to go OS specific, for Linux use `~/.cache/west`, for macOS use `~/Library/Caches/west` (it seems to be the right place but I don't have a macOS setup to check myself). For Windows I think it's fine to use the same as Linux. fish doesn't support Windows "natively", it's only supported through Cygwin, MSYS2 or WSL.

Emplis replied:
Indeed it's a problem if it doesn't work on other OS'es. But I am not sure that using `$__fish_user_data_dir` is the correct way to do. It's not well documented but it seems to be a directory used to load third-party software fish functions.
Here is what the documentation say:
> Directories for other software to put their own functions. These are in the directories under $__fish_user_data_dir (usually ~/.local/share/fish, controlled by the XDG_DATA_HOME environment variable) and in the XDG_DATA_DIRS environment variable, in a subdirectory called fish/vendor_functions.d.

Instead I propose to go OS specific, for Linux use `~/.cache/west`, for macOS use `~/Library/Caches/west` (it seems to be the right place but I don't have a macOS setup to check myself). For Windows I think it's fine to use the same as Linux. fish doesn't support Windows "natively", it's only supported through Cygwin, MSYS2 or WSL.

tejlmand replied:
> But I am not sure that using $__fish_user_data_dir is the correct way to do.

Ref: https://fishshell.com/docs/current/completions.html#where-to-put-completions
Ref: https://github.com/fish-shell/fish-shell/pull/8887

> By default, Fish searches the following for completions, using the first available file that it finds:
> ...
> A user-specified directory for third-party vendor completions, usually ~/.local/share/fish/vendor_completions.d (controlled by the XDG_DATA_HOME environment variable);


So `$__fish_user_data_dir/vendor_completions.d` is actually the right place, as this is related to completion handling.

And then let fish itself decide how it will define `$__fish_user_data_dir` (`XDG_DATA_HOME` on Linux) across OS'es.

Emplis replied:
Sorry I still doesn't agree on that, as I understand those lines, the folder is for the completions themselves, not some "random" files. But at the same time I did not find any example of fish completion that use files.

As a user I appreciate when software use existing "system" directories and doesn't put files everywhere on the system, especially in the home directory.

If that's fine for you, and as long as someone come with examples or some best practices that I may not be aware of, I would like to keep the updated solution that use per OS cache folder.

tejlmand replied:
> the folder is for the completions themselves, not some "random" files.

True, but I kind of see this caching mechanism as part of the completion script.
One thing though, the user may not have added the completion script to that folder in which case it can be argued that we shouldn't place a cache file there.

> I would like to keep the updated solution that use per OS cache folder.

In such case, then `XDG_HOME_DIR` should as minimum be honored on Linux.

Perhaps the best location is under `<west-workspace>/.west/fish/<completion-cache-file>`.
That will place it together with other west related settings for the workspace, and the `<workspace>/.west` folder is always created by `west`.

Ref:
> Local: Settings in this file affect west’s behavior for the current [west workspace](https://docs.zephyrproject.org/latest/glossary.html#term-west-workspace). The file is `.west/config`, relative to the workspace’s root directory.
https://docs.zephyrproject.org/latest/develop/west/config.html#west-configuration-files

Remember, users have the possibility to swap manifest inside a workspace, for example like this:
```
$ west config manifest.path zephyr # Use vanilla Zephyr manifest 
$ west update

$ west config manifest.path nrf # Use nRF Connect SDK manifest 
$ west update
```

Emplis replied:
> In such case, then XDG_HOME_DIR should as minimum be honored on Linux.

You're right, I updated the code

> Perhaps the best location is under <west-workspace>/.west/fish/<completion-cache-file>.
That will place it together with other west related settings for the workspace, and the <workspace>/.west folder is always created by west.

That may be the best but the `.west` folder isn't created by default by west, should fish create it if it doesn't exist?

tejlmand replied:
> That may be the best but the .west folder isn't created by default by west, should fish create it if it doesn't exist?

`.west` folder is created by west when doing a `west init`, and remember that both `west boards` and `west build` are extension commands defined in the west workspace.

Remember, the west workspace folder is not a git folder, it's one level above Zephyr in a standard install.

> If you’ve followed the upstream Zephyr getting started guide, your workspace looks like this:
> ```
> zephyrproject/                 # west topdir
>├── .west/                     # marks the location of the topdir
>│   └── config                 # per-workspace local configuration file
>│
>│   # The manifest repository, never modified by west after creation:
>├── zephyr/                    # .git/ repo
>│   ├── west.yml               # manifest file
>│   └── [... other files ...]
>│
>│   # Projects managed by west:
>├── modules/
>│   └── lib/
>│       └── zcbor/             # .git/ project
>├── net-tools/                 # .git/ project
>└── [ ... other projects ...]
> ```
Ref: https://docs.zephyrproject.org/latest/develop/west/basics.html#example-workspace

Emplis replied:
Oh okay, sorry I misunderstood. In my head you wanted to create a `.west` folder at the manifest path directory.

I updated the code, the structure look like this:
```
zephyrproject/                          
├── .west/                              
│   └── fish                            
│       └── [manifest_path_hash]/       
│           └── fish_boards_completion.cache
├── zephyr/                             
└── [...]                               
```

One small problem with that is that it add a call to `west topdir` but I tested it and it's still acceptable IMO

tejlmand commented on scripts/west_commands/completion/west-completion.fish:
```diff
@@ -196,21 +196,52 @@ function __zephyr_west_complete_help
-    for board in $boards
-        set -l b (string split " " $board)
-        printf "%s\n" $b[1]\t"$b[2]"
+    set -l one_week_in_s 604800
```
One week is a very long time, afaict the cache is not updated even in cases where a new board or variant is added.

Also users switching between upstream and downstream projects, which might have different boards defined, will be hit by the fact that either boards are present, but not completed, or completed but not present in the workspace.

Also this PR doesn't take into consideration the possibility of having multiple workspaces with different releases, for example `zephyr/main` and latest Zephyr LTS.

tejlmand replied:
> west can be called outside a git workspace so it's not possible to rely on it. 

Yes, but `west build` is an extension command and needs to be called from within a `west workspace`, although it can be in called in a folder inside the west workspace that is not within a git repository.

Note, #69904 already updated fish completion from calling list_board.py twice, to just call it once, making it comparable to bash and zsh wrt. to execution time. Bash and zsh already called list_boards only once.

> The solution could be to use west instead but then we need to call west every time making the caching useless.

That would actually be worse, cause `west boards` roughly takes 50% longer than `list_boards.py` due to the overhead of extension command handling.

> Also, if you work on adding a new board there is a high chance that you know it's name

I didn't say anything about being the author of a new board.
It could be a new board or an added variant that you get by a `git fetch` or `west update`.

In such a case the user might know the board name, but not the SoC, CPU clusters, or variants, in which case the completion of those when testing the new board is very helpful. But with a long lived cache, then completion becomes useless.

Emplis replied:
> Yes, but west build is an extension command and needs to be called from within a west workspace, although it can be in called in a folder inside the west workspace that is not within a git repository.

You're absolutely right, I forgot that. I updated the PR to update the cache if the latest git commit of Zephyr is updated.

> Note, https://github.com/zephyrproject-rtos/zephyr/pull/69904 already updated fish completion from calling list_board.py twice, to just call it once, making it comparable to bash and zsh wrt. to execution time. Bash and zsh already called list_boards only once.

I saw that, but to be honest it's still slow, bash and zsh could benefit from this approach.

> That would actually be worse, cause west boards roughly takes 50% longer than list_boards.py due to the overhead of extension command handling.

Yes, calling `west boards` is not an option. To get the latest commit I call `west topdir` and then git.

> I didn't say anything about being the author of a new board.
It could be a new board or an added variant that you get by a git fetch or west update.

> In such a case the user might know the board name, but not the SoC, CPU clusters, or variants, in which case the completion of those when testing the new board is very helpful. But with a long lived cache, then completion becomes useless.

That's right, it should be handled by the new approach based on the latest git commit.

tejlmand reviewed (changes requested):
I like the idea of caching the result, but needs some more rework.

Especially the following must be considered:
- handle jumps between Zephyr releases, as example Zephyr v3.7 (future) and `zephyr/main` using git checkout.
- Multiple workspaces, for example for upstream and downstream work.
   Example could be working in both upstream Zephyr, and a downstream project like [nRF Connect SDK](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/index.html)  where there are different boards defined.

Especially a one week cache is really long in such cases.

Emplis commented:
Rebased to fix merge conflicts.

Emplis commented:
Updated the cache directory to use `~/Library/Caches` on macOS.

Emplis commented:
@tejlmand I updated the cache invalidation trigger. I kept the 7 days but now the cache will also be updated when the latest commit of the Zephyr repository of the active manifest is updated.
The cache will be updated more often but it will solve the problem with the multiple Zephyr version/fork.

tejlmand commented on scripts/west_commands/completion/west-completion.fish:
```diff
@@ -195,15 +195,89 @@ function __zephyr_west_complete_help
+    set -l is_cache_valid 0 # 0: invalid; 1: valid
+
+    set -l cache_folder (__zephyr_west_get_cache_dir)
+    set -l cache_file $cache_folder/fish_boards_completion.cache
+
+    set -l zephyr_sha (git --work-tree (west topdir)/zephyr log -1 --pretty=format:'%H')
```
`west topdir` is faster than a complete `list_boards`, roughly 1/3 of the time compared to list_boards when in tree in Zephyr. The time used by git is insignificant in this context.

However, one risk I see wrt. relying on just the Zephyr repo is that downstream users / teams will often create custom boards as well as having a custom manifest.
And this is where a lot of work will happen and be shared, for example https://github.com/nrfconnect/sdk-nrf, a project which itself is expected to be extended further downstream.

Thus relying on the `(west topdir)/zephyr` will not catch updates to boards specified by other Zephyr modules.
Perhaps a git watch on the manifest repo is a better choice.
This will catch commits to the manifest repo, thus including commits to the manifest file which is changed when the SHA of other projects are updated. This means that a manifest which is updated to point to a different Zephyr revision will also trigger a cache update. Thus board changes in both Zephyr and other projects are handled.

One issue which must be considered in such approach is
Manifest path is not required to be a git repo.
In fact a common use-case is to locally create and manifest file with just a subset of projects.
However, such local manifest folders are usually just containing the manifest file and not code.
So it can be solved an md5 or similar over the manifest file, when manifest path is not a git repo.
Thus `west manifest --path` instead of `west topdir`.
That will also address cases where a manifest is locally modified before a `west update`, when a different revision of a given project is desired.

If using `west manifest --path` in a cache solution, then an `md5sum` over path value (not content) can be used to create a dedicated cache per west workspace.
That would mean the one cache exists for the workspace indicated by `/projects/foo-workspace/repo/west.yml`, and another cache is created for  `/projects/bar-workspace/repo/west.yml`.

Thus both avoiding cluttering caches, as well as allow users to have multiple workspaces / switch manifest using `west config manifest.path foo` and `west config manifest.path bar` within the same workspace.

Emplis replied:
That's a good approach, thanks for the feedback!

I updated the logic to do what you describe. Tell me if I misunderstood something in what you were describing 😀

tejlmand replied:
> I updated the logic to do what you describe. Tell me if I misunderstood something in what you were describing 😀

Thanks, initial look at the logic itself looks good 👍

tejlmand reviewed (commented):
Thanks a lot for the effort, it's highly appreciated 👍 

I believe we're approaching a very good solution where only local modifications may not result in an updated cache. But as long as any `git fetch` / `west update` results in an updated cache when there are changes (remember a Zephyr module may contain extra boards), then I think we have a safe solution.

I have made a small description of such approach.

tejlmand reviewed (changes requested):
Something is not working.

I get the cache file created, but no content:
```
$ west build -p -b nrf52840<TAB>
$ cat ../.west/fish/c03622970e8c56f0abf5936538c5a50b/fish_boards_completion.cache 

$ 
```

Emplis commented:
> Something is not working.
> 
> I get the cache file created, but no content:
> 
> ```
> $ west build -p -b nrf52840<TAB>
> $ cat ../.west/fish/c03622970e8c56f0abf5936538c5a50b/fish_boards_completion.cache 
> 
> $ 
> ```

I am not sure why it wasn't working for you. I updated the logic a bit, saving the hash only if the `west boards` command didn't fail.

I also fixed the boards output, it was wasn't correct.

And I removed the call to `west topdir` and `west manifest --path`, I wrote the equivalent commands in fish, speeding up the completion.

tejlmand commented on scripts/west_commands/completion/west-completion.fish:
```diff
@@ -195,15 +195,141 @@ function __zephyr_west_complete_help
+    set -l cwd_split (string split '/' $cwd)
+
+    while true
+        set -l tmp_path (path normalize /(string join "/" $cwd_split))
```
late observation.

The fish version shipped with Ubuntu 22.04 is 3.3 (and 20.04 ships with 3.1), whereas `path` as built-in was introduced in fish v3.5.
https://fishshell.com/docs/current/relnotes.html#fish-3-5-0-released-june-16-2022

Zephyr supports Ubuntu 20.04 LTS and 22.04 LTS: https://docs.zephyrproject.org/latest/develop/getting_started/index.html#select-and-update-os
> This guide covers Ubuntu version 20.04 LTS and later.

https://packages.ubuntu.com/focal/fish
https://packages.ubuntu.com/jammy/fish

This means some strange errors for fish users on those older releases installing the completion handler.

Emplis replied:
Good catch. This is indeed bad UX, I'll make a PR in the coming weeks to address that issue :slightly_smiling_face:

tejlmand reviewed (approved):
> I am not sure why it wasn't working for you. 

Not sure either.
Have finally had time to come back at testing this a bit more, and seems to be working as intended.

However, I have later discovered that completion doesn't work for Ubuntu 22.04 with its default fish package (v3.3) and therefore also not working on Ubuntu 20.04.

Should not block this improvement, but would be nice to have a followup PR which might give a better user-experience / error message when old fish releases are used.

jori-nordic reviewed (approved):
🐠
